probe/endpoint/conntrack.go

```diff
@@ -0,0 +1,220 @@
+type Layer3 struct {
+	XMLName xml.Name `xml:"layer3"`
+	SrcIP   string   `xml:"src"`
+	DstIP   string   `xml:"dst"`
+}
+
+// Layer4 - these structs are for the parsed conntrack output
+type Layer4 struct {
+	XMLName xml.Name `xml:"layer4"`
+	SrcPort int      `xml:"sport"`
+	DstPort int      `xml:"dport"`
+	Proto   string   `xml:"protoname,attr"`
+}
+
+// Meta - these structs are for the parsed conntrack output
+type Meta struct {
+	XMLName   xml.Name `xml:"meta"`
+	Direction string   `xml:"direction,attr"`
+	Layer3    Layer3   `xml:"layer3"`
+	Layer4    Layer4   `xml:"layer4"`
+	ID        int64    `xml:"id"`
+	State     string   `xml:"state"`
+}
+
+// Flow - these structs are for the parsed conntrack output
+type Flow struct {
+	XMLName xml.Name `xml:"flow"`
+	Metas   []Meta   `xml:"meta"`
+	Type    string   `xml:"type,attr"`
+
+	Original, Reply, Independent *Meta `xml:"-"`
+}
+
+// Conntracker uses the conntrack command to track network connections
+type Conntracker struct {
+	sync.Mutex
+	cmd           exec.Cmd
+	activeFlows   map[int64]Flow // active flows in state != TIME_WAIT
+	bufferedFlows []Flow         // flows coming out of activeFlows spend 1 walk cycle here
+}
+
+// NewConntracker creates and starts a new Conntracter
+func NewConntracker(args ...string) (*Conntracker, error) {
+	if !conntrackModulePresent() {
+		return nil, fmt.Errorf("No conntrack module")
+	}
+	result := &Conntracker{
+		activeFlows: map[int64]Flow{},
+	}
+	go result.run(args...)
+	return result, nil
+}
+
+func conntrackModulePresent() bool {
+	f, err := os.Open(modules)
+	if err != nil {
+		return false
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, conntrackModule) {
+			return true
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		log.Printf("conntrack error: %v", err)
+	}
+
+	log.Printf("conntrack: failed to find module %s", conntrackModule)
+	return false
+}
+
+// NB this is not re-entrant!
+func (c *Conntracker) run(args ...string) {
+	args = append([]string{"-E", "-o", "xml"}, args...)
+	cmd := exec.Command("conntrack", args...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		log.Printf("conntrack error: %v", err)
+		return
+	}
+	if err := cmd.Start(); err != nil {
+		log.Printf("conntrack error: %v", err)
+		return
+	}
+
+	c.Lock()
+	c.cmd = cmd
+	c.Unlock()
+	defer func() {
+		if err := cmd.Wait(); err != nil {
+			log.Printf("conntrack error: %v", err)
+		}
+	}()
+
+	// Swallow the first two lines
+	reader := bufio.NewReader(stdout)
+	if line, err := reader.ReadString('\n'); line != XMLHeader {
+		log.Printf("conntrack invalid output: '%s'", line)
+		return
+	} else if err != nil {
+		log.Printf("conntrack error: %v", err)
+		return
+	}
+	if line, err := reader.ReadString('\n'); line != ConntrackOpenTag {
+		log.Printf("conntrack invalid output: '%s'", line)
+		return
+	} else if err != nil {
+		log.Printf("conntrack error: %v", err)
+		return
+	}
```

Comment:
Should probably switch the order of evaluation for both of these blocks, e.g.

```go
if line, err := reader.ReadString('\n'); err != nil {
    // handle error
} else if line != XMLHeader {
    // handle bad line
}
```

Of course, since your behavior is identical in both cases... not such a big deal.